File: SwordToOffer/code/Test13.java
package ArithmeticCode.SwordToOffer.code;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Hy on 2019/11/29 9:47
 * <p>
 * 调整数组顺序使奇数位于偶数前面
 */
public class Test13 {

    public static class Solution {
        public void reOrderArray(int[] array) {
            int k = 0; //已放好的奇数个数 & 下一个奇数需要放的位置
            //暂存偶数
            List<Integer> evenList = new ArrayList<>();
            //遍历
            for (int i = 0; i < array.length; i++) {
                if ((array[i] & 1) == 1) {
                    //奇数
                    array[k++] = array[i];
                } else {
                    //偶数
                    evenList.add(array[i]);
                }
            }

            //将偶数放回去
            for (int i : evenList) {
                array[k++] = i;
            }
        }

        public void reOrderArray2(int[] array) {
            for (int i = 0; i < array.length; i++) {
                for (int j = array.length - 1; j > i; j--) {
                    if (array[j] % 2 == 1 && array[j - 1] % 2 == 0) {
                        //前偶后奇交换
                        array[j] += array[j - 1];
                        array[j - 1] = array[j] - array[j - 1];
                        array[j] = array[j] - array[j - 1];
                    }
                }
            }
        }
    }


}

File: kotlin/Test13.kt
package ArithmeticCode.kotlin

/**
 * Created by Hy on 2019/11/30 10:13
 *
 * 调整数组顺序使奇数位于偶数前面
 */

fun main() {
    val array = intArrayOf(1, 2, 3, 4, 5)
    Test13().reOrderArray2(array)

    for (num in array) {
        println(num)
    }
}

class Test13 {

    fun reOrderArray(array: IntArray) {
        var k = 0
        val evenArray = ArrayList<Int>()

        for (num in array) {
            if (num and 1 == 1) {
                array[k++] = num
            } else {
                evenArray.add(num)
            }
        }

        for (even in evenArray) {
            array[k++] = even
        }
    }

    fun reOrderArray2(array: IntArray) {
        for (i in array.indices) {
            for (j in 0 until array.size - i - 1) {
                //前偶后奇交换
                if ((array[j] and 1 == 0) && (array[j + 1] and 1 == 1)) {
                    array[j] += array[j + 1]
                    array[j + 1] = array[j] - array[j + 1]
                    array[j] = array[j] - array[j + 1]
                }
            }
        }
    }

}